File: go/service/service_test.go
package service

import (
	"testing"

	"northwind/callbackdirectory/client"
)

type stubClient struct {
	payload client.CountryPayload
}

func (s stubClient) FetchCountry(country string) (client.CountryPayload, error) {
	return s.payload, nil
}

func TestSingleCallingCode(t *testing.T) {
	stub := stubClient{payload: client.CountryPayload{Data: []client.CountryRecord{{Name: "Afghanistan", CallingCodes: []string{"93"}}}}}
	svc := NewDirectoryService(stub)
	result, err := svc.GetPhoneNumbers("Afghanistan", "656445445")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result != "+93 656445445" {
		t.Fatalf("expected +93 656445445, got %s", result)
	}
}

func TestMultipleCallingCodesUsesHighestIndexCode(t *testing.T) {
	stub := stubClient{payload: client.CountryPayload{Data: []client.CountryRecord{{Name: "Puerto Rico", CallingCodes: []string{"1", "1787", "1939"}}}}}
	svc := NewDirectoryService(stub)
	result, err := svc.GetPhoneNumbers("Puerto Rico", "123456789")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result != "+1939 123456789" {
		t.Fatalf("expected +1939 123456789, got %s", result)
	}
}

func TestUnknownCountryReturnsMinusOne(t *testing.T) {
	stub := stubClient{payload: client.CountryPayload{Data: []client.CountryRecord{}}}
	svc := NewDirectoryService(stub)
	result, err := svc.GetPhoneNumbers("Atlantis", "5551234")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result != "-1" {
		t.Fatalf("expected -1, got %s", result)
	}
}

File: javascript/tests/service.test.js
const { createDirectoryService } = require('../src/service');

function stubClient(payload) {
  return { fetchCountry: async () => payload };
}

test('single calling code resolves', async () => {
  const service = createDirectoryService(
    stubClient({ data: [{ name: 'Afghanistan', callingCodes: ['93'] }] })
  );
  const result = await service.getPhoneNumbers('Afghanistan', '656445445');
  expect(result).toBe('+93 656445445');
});

test('multiple calling codes resolve using highest-index code', async () => {
  const service = createDirectoryService(
    stubClient({ data: [{ name: 'Puerto Rico', callingCodes: ['1', '1787', '1939'] }] })
  );
  const result = await service.getPhoneNumbers('Puerto Rico', '123456789');
  expect(result).toBe('+1939 123456789');
});

test('unknown country resolves successfully as -1', async () => {
  const service = createDirectoryService(stubClient({ data: [] }));
  const result = await service.getPhoneNumbers('Atlantis', '5551234');
  expect(result).toBe('-1');
});
